concent_api/core/views.py

```diff
@@ -1241,6 +1209,712 @@ def store_message_and_message_status(
+        subtask__state                   = Subtask.SubtaskState.FORCING_RESULT_TRANSFER.name,  # pylint: disable=no-member
+    )
+
+    for pending_reponse in force_get_task_result_list:
+        report_computed_task    = deserialize_message(pending_reponse.subtask.report_computed_task.data.tobytes())
+        file_transfer_token     = create_file_transfer_token(
+            report_computed_task,
+            client_public_key,
+        )
+        if get_file_status(file_transfer_token):
+            subtask         = pending_reponse.subtask
+            subtask.state   = Subtask.SubtaskState.RESULT_UPLOADED.name  # pylint: disable=no-member
+            subtask.full_clean()
+            subtask.save()
+            mark_message_as_delivered(pending_reponse)
+
+            store_pending_message(
+                response_type       = PendingResponse.ResponseType.ForceGetTaskResultUpload,
+                client_public_key   = subtask.provider.public_key_bytes,
+                queue               = PendingResponse.PendingResponseType.Receive,
+                subtask             = subtask,
+            )
+            store_pending_message(
+                response_type       = PendingResponse.ResponseType.ForceGetTaskResultUpload,
+                client_public_key   = subtask.requestor.public_key_bytes,
+                queue               = PendingResponse.PendingResponseType.Receive,
+                subtask             = subtask,
+            )
+
+
+def update_concent_states(
+    client_public_key: bytes,
+):
+    clients_subtask_list = Subtask.objects.filter(
+        Q(requestor__public_key = b64encode(client_public_key)) | Q(provider__public_key = b64encode(client_public_key)),
+        state__in               = Subtask.SUBTASK_ACTIVE_STATES,
+        next_deadline__lte      = timezone.now()
+    )
+
+    update_file_status(client_public_key)
+    if clients_subtask_list.exists():
+        update_states_of_subtasks(
+            clients_subtask_list,
+        )
+        logging.log_changes_in_subtask_states(
+            b64encode(client_public_key),
+            clients_subtask_list.count(),
+        )
+    else:
+        logging.log_no_changes_in_subtask_states(
+            b64encode(client_public_key)
+        )
+
+
+def update_states_of_subtasks(
+    subtask_list:       list,
+):
+    for subtask in subtask_list:
+        if subtask.state == Subtask.SubtaskState.FORCING_REPORT.name:  # pylint: disable=no-member
+            update_subtask_state(
+                subtask                 = subtask,
+                state                   = Subtask.SubtaskState.REPORTED.name,  # pylint: disable=no-member
+            )
+            store_pending_message(
+                response_type       = PendingResponse.ResponseType.AckReportComputedTask,
+                client_public_key   = subtask.provider.public_key_bytes,
```

Comment:
Ack should be sent to the provider inside `ForceReportComputedTaskResponse`.

Reply:
And it does. Here is just storing message in database. In `handle_messages_from_database` we create new instance of `ForceReportComputedTaskResponse` and send it to client.

Reply:
As I said above - the `ResponseType` value you're using here should be changed then.